client.go

```diff
@@ -153,94 +153,181 @@ func readMyCert(configDir string) (string, string, error) {
+	}
+	info := ConnectInfo{
+		Name: remote,
+		Addr: r.Addr,
+	}
+	if r.Addr[0:5] != "unix:" {
+		certf, keyf, err := ensureMyCert(config.ConfigDir)
+		if err != nil {
+			return nil, err
+		}
+		certBytes, err := ioutil.ReadFile(certf)
+		if err != nil {
+			return nil, err
+		}
+		keyBytes, err := ioutil.ReadFile(keyf)
+		if err != nil {
+			return nil, err
+		}
+		info.ClientPEMCert = string(certBytes)
+		info.ClientPEMKey = string(keyBytes)
+		serverCertPath := config.ServerCertPath(remote)
+		if shared.PathExists(serverCertPath) {
+			cert, err := shared.ReadCert(serverCertPath)
 			if err != nil {
 				return nil, err
 			}
-			c.Certificate = st.Environment.Certificate
+
+			info.ServerPEMCert = string(pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw}))
+		}
+	}
+	c, err := NewClientFromInfo(info)
+	if err != nil {
+		return nil, err
+	}
+	c.Config = *config
+	return c, nil
+}
+
+// ConnectInfo contains the information we need to connect to a specific LXD server
+type ConnectInfo struct {
+	// Name is a simple identifier for the remote server. In 'lxc' it is
+	// the name used to lookup the address and other information in the
+	// config.yml file.
+	Name string
+	// Addr is the host address to connect to. It can be unix: to indicate
+	// we should connect over a unix socket, or it can be an IP Address or
+	// Hostname, or an https:// URL.
+	Addr string
+	// ClientPEMCert is the PEM encoded bytes of the client's certificate.
+	// If Addr indicates a Unix socket, the certificate and key bytes will
+	// not be used.
+	ClientPEMCert string
+	// ClientPEMKey is the PEM encoded private bytes of the client's key associated with its certificate
+	ClientPEMKey string
+	// ServerPEMCert is the PEM encoded server certificate that we are
+	// connecting to. It can be the empty string if we do not know the
+	// server's certificate yet.
+	ServerPEMCert string
+}
+
+// how to handle ServerCerts that we want to cache?
+
+func connectViaUnix(c *Client, addr string) error {
+	if addr == "unix://" {
+		addr = fmt.Sprintf("unix:%s", shared.VarPath("unix.socket"))
+	}
+
+	c.BaseURL = "http://unix.socket"
+	c.BaseWSURL = "ws://unix.socket"
+	c.Transport = "unix"
+	r := &RemoteConfig{
+		Addr: addr,
+		// TODO: (jam) RemoteConfig.Public is not relevant for the purposes of
+		// an active connection. It only seems to be used to display server
+		// information in "list", though if we aren't getting that information
+		// from the service itself, it seems like it can't be guaranteed to be
+		// correct, as it is just the local information about a remote server.
+		// Is there another reason it is here?
+		Public: false,
+	}
+	uDial := func(networ, addr string) (net.Conn, error) {
+		// TODO: (jam) Why are we ignoring the passed in 'addr' and only using our own Addr?
+		// Do we allow other code to mutate Client.Remote.Addr and
+		// change what we connect to, and still ignore the 'adrd'
+		// passed to Dial?
```

Comment:
So I spent some time thinking about this, and given the function is called Dial, I have the feeling it will be turning a URL back into a host:port sort of address. Which isn't what we want for a local unix socket. I did some testing and found that was exactly right. addr was "unix.socket:80". So I updated this comment to explain why we are ignoring it.